colossalai/tensor/colo_tensor.py

```diff
@@ -127,85 +132,21 @@ def torch_tensor(self) -> torch.Tensor:
-        if shard == True:
-            self.shard()
-
-    def set_shard_pattern(self, shard_pattern: ShardPattern):
-        self._shard_pattern = shard_pattern
-
-    def shard(self):
-        assert self._shard_spec is not None, 'You should call set_spec() before _shard() ColoTensor.'
-        if self._shard_pattern is not ShardPattern.NA:    # reshard
-            self.gather()
-        # Model Parameters
-        if self._shard_spec.num_action == 1:
-            parallel_action = self._shard_spec.get_action_by_compute_pattern(self._shard_spec.compute_patterns[0])
-            if parallel_action.compute_pattern in [
-                    ComputePattern.TP1DRow_Linear, ComputePattern.TP1DCol_Embedding, ComputePattern.TP1DCol_mm
-            ]:
-                self._shard_1d(parallel_action=parallel_action, dim=-1)
-                # We bind our ComputePattern on weight, which has to be transposed when linear().
-                self._shard_pattern = ShardPattern.Col
-            elif parallel_action.compute_pattern in [
-                    ComputePattern.TP1DCol_Linear, ComputePattern.TP1DRow_Embedding, ComputePattern.TP1DRow_mm
-            ]:
-                self._shard_1d(parallel_action=parallel_action, dim=0)
-                self._shard_pattern = ShardPattern.Row
-            else:
-                raise NotImplementedError
-
-    def gather(self):
-        assert not self.is_model_data(), 'Currently we only support gather Activation ColoTensor.'
-        assert not self.is_gathered(), 'Only sharded ColoTensor can be gathered.'
-        parallel_action = self._shard_spec.get_action_by_compute_pattern(ComputePattern.DP)
-        dim = self._get_gather_dim()
-        self._torch_tensor = gather_forward_split_backward(self._torch_tensor, parallel_action.parallel_mode, dim=dim)
-        self._shard_pattern = ShardPattern.NA
-        self._size = self._torch_tensor.size()
-
-    def global_torch_tensor(self) -> torch.Tensor:
-        out_tensor = self.torch_tensor()
-        if self.is_gathered():
-            return out_tensor
-
-        parallel_action = self._shard_spec.get_action_by_compute_pattern(ComputePattern.DP)
-        world_size = gpc.get_world_size(parallel_action.parallel_mode)
-        if world_size == 1:
-            return out_tensor
-
-        rank = gpc.get_local_rank(parallel_action.parallel_mode)
-        tensor_list = [torch.empty_like(out_tensor) for _ in range(world_size)]
-        tensor_list[rank] = out_tensor
-        torch.distributed.all_gather(tensor_list, out_tensor, group=gpc.get_group(parallel_action.parallel_mode))
-
-        dim = self._get_gather_dim()
-        out_tensor = torch.cat(tensor_list, dim=dim).contiguous()
-
-        return out_tensor
-
-    def is_gathered(self) -> bool:
-        return self._shard_pattern == ShardPattern.NA
+    def set_spec(self, spec: TensorSpec, shard: bool = True, init_pg: bool = False) -> None:
+        spec = copy(spec)
+        if init_pg:
```

Comment:
What is init_pg?

Reply:
Some tensor is initialized with wrong process group. We cannot handle process group change now, so we just copy new process group to the previous DistSpec. It's a temporary design and I may change this function in the future.

Reply:
I already removed this arg.

Reply:
ok